### PR TITLE
add bazel-remote as a go package and a nixos module

### PR DESCRIPTION
**Pull request by shmish111**
###### Motivation for this change

Add https://github.com/buchgr/bazel-remote as a NixOS module

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/misc/bazel-remote.nix [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    systemd.services.bazel-remote =
+      { description = "Bazel Remote Cache";
+
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/bazel-remote ${bazelRemoteConfig}";
+          User = cfg.user;
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: Try using `DynamicUser = true;`, then there's no need to create a user

File: nixos/modules/services/misc/bazel-remote.nix [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+      user = mkOption {
+        type = types.str;
+        default = "bazel";
+        description = "User account under which Bazel Remote runs.";
+      };
+
+      pidFile = mkOption {
+        type = types.path;
+        default = "/var/lib/bazelRemote/bazelRemote.pid";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: I don't think there's a need to have this as an option, no reason to ever want to change this.

File: nixos/modules/services/misc/bazel-remote.nix [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        description = "";
+      };
+
+      port = mkOption {
+        type = types.int;
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: `types.port`

File: nixos/modules/services/misc/bazel-remote.nix [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        default = false;
+        description = ''
+          Whether to open ports in the firewall for the server.
+        '';
+      };
+
+      bind = mkOption {
+        type = with types; nullOr str;
+        default = null; # 127.0.0.1
+        description = "The IP interface to bind to.";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: What does `null` represent? Should be in the description.

File: nixos/modules/services/misc/bazel-remote.nix [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        example = "127.0.0.1";
+      };
+
+      yamlConfig = mkOption {
+        type = with types; nullOr lines;
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: If it's a path then it should be `nullOr path`

File: nixos/modules/services/misc/bazel-remote.nix [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      };
+
+      bind = mkOption {
+        type = with types; nullOr str;
+        default = null; # 127.0.0.1
+        description = "The IP interface to bind to.";
+        example = "127.0.0.1";
+      };
+
+      yamlConfig = mkOption {
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: If it's a path, this should be named `configFile`

File: nixos/modules/services/misc/bazel-remote.nix [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+      };
+
+      dir = mkOption {
+        type = types.path;
+        description = "Directory path where to store the cache contents";
+      };
+
+      maxSize = mkOption {
+        type = types.int;
+        description = "The maximum size of the remote cache in GiB";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: Does it make sense to add a default here?

File: nixos/modules/services/misc/bazel-remote.nix [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+      yamlConfig = mkOption {
+        type = with types; nullOr lines;
+        default = null;
+        description = "Path to a YAML configuration file. If this option is specified then all other bazel-remote options are ignored";
+      };
+
+      dir = mkOption {
+        type = types.path;
+        description = "Directory path where to store the cache contents";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: Does it make sense to have a default here?

File: nixos/modules/services/misc/bazel-remote.nix [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+      maxSize = mkOption {
+        type = types.int;
+        description = "The maximum size of the remote cache in GiB";
+      };
+
+      htpasswdFile = mkOption {
+        type = with types; nullOr path;
+        default = null;
+        description = "Path to a .htpasswd file. This flag is optional. Please read https://httpd.apache.org/docs/2.4/programs/htpasswd.html";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: Use this to get links to render correctly: `<link xlink:href="https://example.com"/>`, and encase files in `<filepath>...</filepath>`. Also there's no need to point out that this setting is optional.

File: nixos/modules/services/misc/bazel-remote.nix [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+      };
+
+      tlsCertFile = mkOption {
+        type = with types; nullOr path;
+        default = null;
+        description = "Path to a pem encoded certificate file";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: All descriptions should end with a colon (for other descriptions as well).

File: nixos/modules/services/misc/bazel-remote.nix [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+      tlsKeyFile = mkOption {
+        type = with types; nullOr path;
+        default = null;
+        description = "Path to a pem encoded certificate file";
+      };
+
+      idleTimeout = mkOption {
+        type = types.int;
+        default = 0;
+        description = "The maximum period in seconds of having received no request after which the server will shut itself down. Disabled by default";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: You should indicate that `0` stands for it being disabled.

File: pkgs/servers/bazel-remote/default.nix [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  };
+
+  goPackagePath = "github.com/buchgr/bazel-remote";
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/buchgr/bazel-remote;
+    description = "A remote HTTP/1.1 cache for Bazel https://bazel.build";
+    license = licenses.asl20;
+    maintainers = [ maintainers.shmish111 ];
+    platforms = platforms.all;
```
Comment (Infinisil) [inline pkgs/servers/bazel-remote/default.nix]: No need to add platforms, `buildGoPackage` does that automatically.

File: pkgs/servers/bazel-remote/default.nix [hunk 13]
```diff
@@ -0,0 +1,25 @@
+{ stdenv, fetchFromGitHub, buildGoPackage, cacert, git, go }:
```
Comment (Infinisil) [inline pkgs/servers/bazel-remote/default.nix]: You're not using `cacert`, `git`, or `go`

File: nixos/modules/services/misc/bazel-remote.nix [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether to enable the Bazel Remote server.";
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.bazel-remote;
+        defaultText = "pkgs.bazelRemote";
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: That's not right

File: nixos/modules/services/misc/bazel-remote.nix [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+  ###### interface
+
+  options = {
+
+    services.bazelRemote = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
```
Comment (Infinisil) [inline nixos/modules/services/misc/bazel-remote.nix]: You can use `mkEnableOption`

**Review by infinisil — CHANGES_REQUESTED** [review]
In addition to my comments, you need to split this PR into three commits in this order:
- One for adding yourself as a maintainer
- One for adding the package
- And one for adding the NixOS module

And format your commit message according to the [Contribution Guidelines](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes), also linked to from the PR template.

File: nixos/modules/services/misc/bazel-remote.nix (hunk 14 above)
Comment (shmish111) [reply nixos/modules/services/misc/bazel-remote.nix]: What should this be? It was a copy and paste from another module (can't remember which one)

File: nixos/modules/services/misc/bazel-remote.nix (hunk 2 above)
Comment (shmish111) [reply nixos/modules/services/misc/bazel-remote.nix]: OK, I'll remove it, again from the copy and paste

File: nixos/modules/services/misc/bazel-remote.nix (hunk 14 above)
Comment (Infinisil) [reply nixos/modules/services/misc/bazel-remote.nix]: pkgs.bazelRemote doesn't exist

File: nixos/modules/services/misc/bazel-remote.nix (hunk 8 above)
Comment (shmish111) [reply nixos/modules/services/misc/bazel-remote.nix]: bazel-remote doesn't, I copied all the options and descriptions from the `bazel-remote --help` and `dir` and `maxSize` are required their so I'm leaving them as required here.

File: nixos/modules/services/misc/bazel-remote.nix (hunk 10 above)
Comment (shmish111) [reply nixos/modules/services/misc/bazel-remote.nix]: A colon? `:`? Why is that?

**Comment by shmish111** [comment]
Thanks @Infinisil I'll split it up but I just want to wait for your answers to my questions first.

File: nixos/modules/services/misc/bazel-remote.nix (hunk 10 above)
Comment (Infinisil) [reply nixos/modules/services/misc/bazel-remote.nix]: Oh sorry, confused it with a dot/full stop. All descriptions should have a `.` at the end.

File: nixos/modules/services/misc/bazel-remote.nix (hunk 8 above)
Comment (Infinisil) [reply nixos/modules/services/misc/bazel-remote.nix]: Alright that's fine then.

**Comment by Infinisil** [comment]
I said to split it into three *commits*, not pull requests. Having PR's that depend on each other isn't very nice. But it doesn't matter now

**Pull request closed without merging**